src/components/cylc/tree/tree.js

```diff
@@ -0,0 +1,497 @@
+          parent.children.push(familyProxy)
+        }
+      }
+    }
+  }
+
+  /**
+   * @param {null|{
+   *   id: string,
+   *   node: Object,
+   *   children: []
+   * }} familyProxy
+   */
+  updateFamilyProxy (familyProxy) {
+    if (familyProxy) {
+      const node = this.lookup.get(familyProxy.id)
+      if (node) {
+        merge(node, familyProxy)
+        if (!node.node.state) {
+          node.node.state = ''
+        }
+      }
+    }
+  }
+
+  /**
+   * @param {null|string} familyProxyId
+   */
+  removeFamilyProxy (familyProxyId) {
+    let node
+    let nodeId
+    let parentId
+    // NOTE: when deleting the root family, we can also remove the entire cycle point
+    if (familyProxyId.endsWith('|root')) {
+      // 0 has the owner, 1 has the workflow Id, 2 has the cycle point, and 3 the family name
+      const [owner, workflowId, cyclePoint] = familyProxyId.split('|')
+      nodeId = cyclePoint
+      node = this.lookup.get(nodeId)
+      parentId = `${owner}|${workflowId}`
+    } else {
+      nodeId = familyProxyId
+      node = this.lookup.get(nodeId)
+      if (node && node.node && node.node.firstParent) {
+        if (node.node.firstParent.name === FAMILY_ROOT) {
+          parentId = node.node.cyclePoint
+        } else {
+          parentId = node.node.firstParent.id
+        }
+      }
+    }
+    if (node) {
+      this.recursivelyRemoveNode(node)
+      const parent = this.lookup.get(parentId)
+      // If the parent has already been removed from the lookup map, there won't be any parent here
+      if (parent) {
+        parent.children.splice(parent.children.indexOf(node), 1)
+      }
+    }
+  }
+
+  // --- Task proxies
+
+  /**
+   * @param {null|{
+   *   id: string,
+   *   node: Object,
+   *   children: []
+   * }} taskProxy
+   */
+  addTaskProxy (taskProxy) {
+    if (taskProxy && !this.lookup.has(taskProxy.id)) {
+      // progress starts at 0
+      taskProxy.node.progress = 0
+      this.lookup.set(taskProxy.id, taskProxy)
+      let parent
+      if (taskProxy.node.firstParent.name === FAMILY_ROOT) {
+        // if the parent is root, we must instead attach this node to the cyclepoint!
+        parent = this.lookup.get(taskProxy.node.cyclePoint)
+      } else {
+        // otherwise its parent is another family proxy node and **MUST** already exist
+        parent = this.lookup.get(taskProxy.node.firstParent.id)
+      }
+      if (!parent) {
+        // eslint-disable-next-line no-console
+        console.error(`Missing parent ${taskProxy.node.firstParent.id}`)
+      } else {
+        parent.children.push(taskProxy)
+      }
+    }
+  }
+
+  /**
+   * @param {null|{
+   *   id: string,
+   *   node: Object,
+   *   children: []
+   * }} taskProxy
+   */
+  updateTaskProxy (taskProxy) {
+    if (taskProxy) {
+      const node = this.lookup.get(taskProxy.id)
+      if (node) {
+        computeTaskProgress(taskProxy)
+        merge(node, taskProxy)
```

Comment:
Could the `merge` overwrite the progress computed by `computeTaskProgress`?

Reply:
I don't think so. Here we are updating a task proxy. The `node` is an existing task proxy, that may be displayed in the UI too.

What we are doing, is given the new data for the task proxy in `deltas.updated`, we compute its progress (as UIS doesn't provide that info, but gives us the `meanElapsedTime`, `startedTime`, etc).

And then we `merge(object, sources...)`, where the left side object is the one receiving updates.

So I think `node` will get updated with the most recent progress from the `deltas.updated` data.